File: MNPC 2023/Problem K. Maze Analysis/solution.cpp
#include <bits/stdc++.h>
using namespace std;

struct Trie {
    Trie* left;   
    Trie* right;  
    int count;

    Trie() {
        left = nullptr;
        right = nullptr;
        count = 0;
    }

    void insert(const string& s) {
        Trie* root = this;
        for (char c : s) {
            if (c == 'L') {
                if (!root->left) {
                    root->left = new Trie();
                }
                root = root->left;
            } else {
                if (!root->right) {
                    root->right = new Trie();
                }
                root = root->right;
            }
            root->count++;
        }
    }

    int getCount(const string& s) {
        Trie* root = this;
        for (char c : s) {
            if (c == 'L') {
                if (!root->left) return 0;
                root = root->left;
            } else {
                if (!root->right) return 0;
                root = root->right;
            }
        }
        return root->count;
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    Trie t;
    int n, k;
    cin >> n >> k;

    for (int i = 0; i < n; ++i) {
        string s; cin >> s; 
        t.insert(s);
    }

    for (int i = 0; i < k; ++i) {
        string s; cin >> s ;
        cout << t.getCount(s) << "\n";
    }

    return 0;
}

File: MNPC 2023/Problem K. Maze Analysis/Solution.java
import java.util.*;

public class Solution {
    static class Trie {
        Trie left  ; 
        Trie right ; 
        int count ; 
        public Trie(){
            this.left = null ; 
            this.right = null ; 
            this.count = 0 ; 
        }
        public void insert( String s ){
            Trie root = this ;  
            for( char c : s.toCharArray()){
                if( c == 'L' ){
                    if( root.left == null ){
                        root.left = new Trie() ; 
                    }
                    root = root.left; 
                }else {
                    if( root.right == null ){
                        root.right = new Trie() ; 
                    }
                    root = root.right; 
                }
                root.count++ ; 
            }
        }
        public int getCount( String s ){
            Trie root = this ;  
            for( char c : s.toCharArray() ){
                if( c == 'L' ){
                    if( root.left == null ) return 0 ; 
                    root = root.left ; 
                }else {
                    if( root.right == null ) return 0 ; 
                    root = root.right ;  
                }
            }
            return root.count ; 
        }
    }
    public static void main( String[] args ){
        Scanner sc = new Scanner( System.in) ; 
        Trie t = new Trie() ; 
        int n = sc.nextInt() ; 
        int k = sc.nextInt() ; 
        sc.nextLine() ;
        for( int i=0 ; i<n ; ++i ){
            t.insert(sc.nextLine());
        } 
        for( int i=0 ; i<k ; ++i ){
            System.out.println( t.getCount(sc.nextLine()) ) ; 
        }
    }
}